### DISCUSSION TITLE
Support automatic permission elevation using PolKit.

### DISCUSSION THREAD

**Discussion by roke-julian-lockhart** [hardinfo2/hardinfo2#59] [Ideas]
Instead of displaying the undermentioned:

> ![Screenshot_20240812_203746](https://github.com/user-attachments/assets/8a84816d-bd09-47c4-a1b0-a72dd031fd2d)

HardInfo2 should automatically prompt the user to elevate its permissions, like KDE Partition Manager does, using PolKit:

<details>

1.	> ![Screenshot_20240719_160828](https://github.com/user-attachments/assets/242ffe52-b391-4674-be49-29b3c76e79eb)
1.	> ![Screenshot_20240719_160819](https://github.com/user-attachments/assets/fa6e0e89-17a1-4aaa-9bdc-dcf295c5e0a2)

</details>

This is especially important because `sudo hardinfo` doesn't work well for me:

<details>

1.	> ![Screenshot_20240715_141754](https://github.com/user-attachments/assets/40d20551-09c6-47da-810a-5a57b5228a70)
1.	> ![Screenshot_20240715_141752](https://github.com/user-attachments/assets/26dbf897-d579-4f0f-822e-6778065ee834)
1.	> ![Screenshot_20240715_141748](https://github.com/user-attachments/assets/168b71a1-2dae-4515-9092-14f64cf3edf8)

</details>

**Comment by hwspeedy** [comment]
Thanx for suggestions.

**1. root access**
Priority is to only need user access - all we do is displaying information to user, that should be made possible through distro.
I consider polkit an unmature daemon - but that could have changed - still have 130 issues, so maybe not...
As an user, I prefer info about why any program needs root access over just doing something.
Also remember that this program can be used in CLI mode also.

I have been changing the source code from needing root access to explain exactly what is needed, so that a user can add these mods to their startup script. 
Eg. Memory SPD info needs the eeprom driver and some distros still have no user read access to dmi information - so that is explained with suggestions to fix the distro.
If possible to get some info as user, then present that and tell that root access can give more info.

**2. problem description**
I have not seen the problems, you see - please improve description of each problem for reproduction and including severity description for you.

**Comment by roke-julian-lockhart** [comment]
> https://github.com/hardinfo2/hardinfo2/issues/54#issuecomment-2228740533

@hwspeedy:

1.	> Priority is to only need user access - all we do is displaying information to user, that should be made possible through distro. I consider polkit an unmature daemon - but that could have changed - still have 130 issues, so maybe not...

	As does the Kernel, and SystemD. That's surely not an indicator of immaturity - the opposite.

1.	> As an user, I prefer info about why any program needs root access over just doing something.

	See my new screenshot - that information is provided:

	> ![Screenshot_20240719_160819](https://github.com/user-attachments/assets/6f66c6b6-3426-41dd-87f3-68206c94f6ed)

1.	> Also remember that this program can be used in CLI mode also.

	`snapd`'s `snap` could be used as an example. ~~It uses a text prompt in CLI mode. Otherwise, it utilizes PolKit's GUI. PolKit might provide both UIs, but I don't know.~~ `sudo` could always be utilized in that case, since it's designed for CLIs:

	1.	If I invoke:

		```sh
		snap refresh
		```

		...the undermentioned appears:
	
		> ![Screenshot_20240719_161720](https://github.com/user-attachments/assets/68bf7096-d769-4e88-a574-254d5986034b)

	1.	If I press <kbd>Cancel</kbd>:
	
		> ```log
		> PS /home/RokeJulianLockhart> snap refresh  
		> error: access denied (try with sudo)
		> ```

1.	> Memory SPD info needs the eeprom driver and some distros still have no user read access to dmi information - so that is explained with suggestions to fix the distro.

	Do you mean that someone OSes don't package the necessary dependencies?

1.	> I have not seen the problems, you see - please improve description of each problem for reproduction and including severity description for you.

	I don't know what you mean by that. Please elaborate.

**Comment by hwspeedy** [comment]
Hi @RokeJulianLockhart 

Thanx for you input - I will convert this issue to a discussion, where you and other can continue.

Mature means that is solves all problems, does not introduce new once and work for the last 10 years (Timeframe of most LTS). 

1. I do not see the problems you see - hence no problem to solve.
2. Polkit is unmature and the right direction is for the remaining sudo needing information is to pass along the request for changes and wait for distro and D-Bus service apps updates or tell the user how to do the fix. No need to complicate nor introduce unsafe quirks.

**Reply by roke-julian-lockhart** [reply]
> I do not see the problems you see - hence no problem to solve.

As in, when you run `sudo hardinfo2`, those dialogues don't appear? If so, what OS are you using, for comparison? You should see it on https://download.fedoraproject.org/pub/fedora/linux/releases/40/Spins/x86_64/iso/Fedora-KDE-Live-x86_64-40-1.14.iso using https://kojipkgs.fedoraproject.org//packages/hardinfo2/2.1.11/1.fc40/x86_64/hardinfo2-2.1.11-1.fc40.x86_64.rpm (from https://koji.fedoraproject.org/koji/rpminfo?rpmID=38833174).

> Polkit is unmature and the right direction is for the remaining sudo needing information is to pass along the request for changes and wait for distro and D-Bus service apps updates or tell the user how to do the fix. No need to complicate nor introduce unsafe quirks.

I'm really surprised that you consider it to be such, because all of [KDE](https://security.opensuse.org/2024/04/02/kde6-dbus-polkit.html#polkit-overview), GNOME, and Fedora's software is being ported to it.

**Reply by hwspeedy** [reply]
I test on the 30+ distros - I do not see the warnings dialogs from your details section above - in any of them.
See https://hardinfo2.org/github?latest_prerelease for list of distros.

Sorry but there are no real polkit rules supplied in fedora41/debian12 - so still experimental.
I see the move to be services - but no clear direction/solution/problem description... - Anyway we should not need root access for showing system info to the user.

**Reply by roke-julian-lockhart** [reply]
> I test on the 30+ distros - I do not see the warnings dialogs from your details section above - in any of them.

@hwspeedy, have you tested https://download.fedoraproject.org/pub/fedora/linux/releases/40/Spins/x86_64/iso/Fedora-KDE-Live-x86_64-40-1.14.iso?

> Anyway we should not need root access for showing system info to the user.

Commands like `lshw` and `dmidecode` require superuser access for reasonable reasons that would require seriously modifying the OS to remediate. I'm not sure what you suggest otherwise.

**Comment by hwspeedy** [comment]
**ATK warnings**
yes kde on fedora 40 is tested - no problems.
The warning messages that you see, seams to be due to the ATK, which is transforming from external GTK3 to internal GTK4 and in that process has warnings enabled and perhaps errors...
https://stackoverflow.com/questions/75406844/not-loading-module-atk-bridge-the-functionality-is-provided-by-gtk-natively

**Regarding root**
The goal is to not use root access for the program and still show all information. 
Current state - root access gives:
1. more memory information - dmitables memory modules section (distro)
2. addresses on resources (kernel)

None of these are critical to the program and works fine with sudo.
Nor would it be critical for the kernel/distro to change rights for user read.

**Reply by hwspeedy** [reply]
Added oneshot service for hardinfo2:
1. ioports and iomem base addresses 
2. dmi tables read rights.
3. spd module loading 

https://github.com/hardinfo2/hardinfo2/commit/0d12b5c8e7b33e1b74e8e151d12e6fa071e3c1b6

**Security Check**
Can anyone see any security concerns if the user can see these base addresses?
Can anyone see any security concerns if the user can see dmitables?
Can anyone see any security concerns if the spd modules are loaded by distro?
